### ISSUE TITLE
t.c.telnet.ITelnetProtocol.unhandledSubnegotiation doesn't reflect reality

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4751]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)| @bitglue reported|
|-|-|
|Trac ID|trac#4751|
|Type|defect|
|Created|2010-12-01 22:39:29Z|

Caught this with z.i.verify.verifyObject, when testing my own subclass of TelnetProtocol:

    raise BrokenMethodImplementation(name, mess)
zope.interface.exceptions.BrokenMethodImplementation: The implementation of unhandledSubnegotiation violates its contract because implementation requires too many arguments.

Looking through the source, looks like the interface is missing the 'command' argument.

Attachments:

* [telnet.ITelnetProtocol-4751.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2b1/2b19bb77b47f2394e3b956bd8611c68adde7c264/d24e4b7b742591a1b4c43bc405ebaf8e050f82a5.patch) (1050 bytes) - added by indigo on 2011-09-30 18:57:46Z - 
* [telnet.ITelnetProtocol-4751.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2b1/2b19bb77b47f2394e3b956bd8611c68adde7c264/f63c99ff7f4aaf17c6491f4c1b3b6ae7fbb76c59.patch) (3320 bytes) - added by indigo on 2011-09-30 20:47:26Z - addressing review feedback

<details><summary>Searchable metadata</summary>

```
trac-id__4751 4751
type__defect defect
reporter__indigo indigo
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__telnet_easy telnet easy
time__1291243169000000 1291243169000000
changetime__1317456708000000 1317456708000000
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @bitglue|
|-|-|

Thanks!

  1. Ideally, the arguments would be documented in the interface method docstring.  Feel like going that extra step?
  1. The test method should have a docstring.  Also, there's sort of a convention to name that test `test_interface`.  Not a hard requirement, but I think it looks better than all those upper case letters. ;)

This is probably worth a bugfix news fragment.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_4"></a>|@bitglue removed owner|
|-|-|

Arguments documented. Correctly, I hope. Test method renamed and docstringed. News fragment added.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @glyph|
|-|-|

Looks great.  Thanks for adding all those docs.

I will land momentarily.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set status to `closed`|
|-|-|

(In [32724]) Add documentation and correct interface parameters for ITelnetProtocol.

Author: indigo

Reviewer: glyph

Fixes: [#4751](https://github.com/twisted/twisted/issues/4751)

The interface for ITelnetProtocol was previously mis-specified, and this change
corrects it in addition to adding @param and @type information for many methods
on that interface.
